File: algorithms/lab00/python/task-1/sol2_heap.py
# SPDX-License-Identifier: BSD-3-Clause

import heapq

class Solution:
    # Solutie: Folosim un min heap pentru a stoca cele mai mari k elemente.
    # Complexitate: O(nlogk).
    @staticmethod
    def find_kth_largest(nums: list[int], k: int) -> int:
        # pq este un min heap.
        heap: list[int] = nums[:k]
        heapq.heapify(heap)
        # Insereaza primele k elemente in min heap.
        # Insereaza elementele ramase in min heap.
        for x in nums[k:]:
            heapq.heappushpop(heap, x)
        # Elementul din varful min heap-ului este al k-lea cel mai mare element.
        return heap[0]

File: algorithms/lab00/rust/task-1/sol2_heap.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::cmp::Reverse;
use std::collections::BinaryHeap;

// Solutie: Folosim un min heap pentru a stoca cele mai mari k elemente.
// Complexitate: O(n log k)

impl Solution {
    pub fn find_kth_largest(nums: &[i32], k: i32) -> i32 {
        let k = k as usize;
        let mut heap: BinaryHeap<Reverse<i32>> = BinaryHeap::with_capacity(k);

        for num in nums {
            heap.push(Reverse(num));
            if heap.len() > k {
                heap.pop(); // scoate cel mai mic dintre cele k+1
            }
        }

        heap.peek().unwrap().0
    }
}
